cmus.c

```diff
@@ -104,6 +105,38 @@ void cmus_prev(void)
 		player_set_file(info);
 }
 
+void cmus_next_album(void)
+{
+	struct track_info *info;
+
+	view_clear(QUEUE_VIEW);
+
+	if (play_library) {
+		info = lib_goto_next_album();
+	} else {
+		info = pl_goto_next();
+	}
+
+	if (info)
+		player_set_file(info);
+}
+
+void cmus_prev_album(void)
+{
+	struct track_info *info;
+
+	view_clear(QUEUE_VIEW);
```

Comment:
I'm on the fence about this one (and the one in `cmus_next_album`).

Currently, switching tracks manually while there are tracks in the queue will replace the current track and the overall playback position, while preserving the playlist. Going to the next track will pop an item from the queue, and going to the previous track is the same as manually selecting that track. After the playlist ends, playback will resume at the previous location (i.e. the last track selected by something other than the playlist).

While this change makes sense in some ways, it is not completely consistent with the existing commands. If this is kept, a note should probably be added to the documentation.

@gavtroy, can you explain your reasoning behind doing this?

Reply:
To me "next album" doesn't even make sense while tracks are queued, but obviously the user can do it so we have to think about it. The most intuitive behavior would be to skip through the queue until finding a track with a different album, but that would be painful to implement for such a rare scenario, especially considering the queue contains "simple" tracks (i.e. not album aware). So I considered what easier options I had:
1. Ignore the command.
2. Go to next queue item even if it's the same album.
3. Go to 1st track of album outside the queue, but (surprisingly?) continue with the queue immediately afterwards.
4. Clear queue and go to 1st track of album outside the queue.

These options all looked bad so I just went with what I personally thought would be least surprising to the user (4).

For `prev_album` it's even less clear what the user wants to happen. Since I had decided to clear the queue in `next_album` I supposed the same reasoning would apply here, although you're right that it is then inconsistent with `cmus_prev`.

I don't have strong feelings about this either way.

Reply:
> The most intuitive behavior would be to skip through the queue until finding a track with a different album

Personally, I'd be against that because of the complexity (as you mentioned) and also since that's different than how track shuffle works (it doesn't pick a random track from the queue).

> Ignore the command.

I agree that that's unacceptable.

> Go to next queue item even if it's the same album.

That's what the existing next/prev commands are for, so I agree with not going with that either.

> Go to 1st track of album outside the queue, but (surprisingly?) continue with the queue immediately afterwards

That's more of what I'd have expected based on the existing behavior, which I admit is surprising at first (but I've gotten quite used to it and I like it).

> Clear queue and go to 1st track of album outside the queue.
>
> These options all looked bad so I just went with what I personally thought would be least surprising to the user (4).

I agree with the logic of that compared to other players, but I still think it's different from the existing behavior. OTOH, it's a new feature, so I guess it doesn't *have* to match the other commands.

> For prev_album it's even less clear what the user wants to happen

Personally, I believe, it should go to the first track of the previous album or whatever the shuffling is done by (like it does now), but keep the queue (like the current prev).

> I don't have strong feelings about this either way.

Neither do I, and I generally like it the way it is other than the somewhat unexpected queue behavior. Personally, I'd prefer to preserve the queue, but I'd be fine with either adding a note about it to the documentation. The way I see it, the queue should only be affected by commands which manipulate it or pop from it, and should otherwise be ignored by the other commands.

It'd be nice to have another opinion from someone else, though.

Reply:
I've decided to just remove the clear for now. Other opinions are still welcome if anyone feels strongly about it...